### PR TITLE
Cstd ptrs

### PR DESCRIPTION
**Pull request by poshul**
Adds an option to switch which flavour of shared_ptrs to use.

### CODE REVIEW

**Comment by uweschmitt** [comment]
Following tests are failing using Python `3.7.2` and Cython `0.29.13`:

```
tests/test_code_generator.py::test_shared_ptr FAILED                                                                                                                                                     [ 25%]
tests/test_code_generator.py::test_libcpp FAILED                                                                                                                                                         [ 50%]
tests/test_code_generator.py::test_stl_libcpp FAILED                                                                                                                                                     [ 75%]
tests/test_main.py::test_run FAILED
```

**Comment by uweschmitt** [comment]
@poshul : Thanks for your contribution.

I reviewed you commits and have a few comments: 
- please split the pull request into different parts, we are mixing here different tasks which should be also split as pull requests. Else I have to dissect your commits for the parts I'm willing to accept and to those where I have concerns.
- we have a discussion at https://github.com/uweschmitt/autowrap/pull/82 to work on the https://github.com/uweschmitt/autowrap/pull/83/commits/d81ef2af07c601e959741d81ed0a1b00068d321f change
- The configuration option to use boost  shared pointers to use is actually about supporting C++11 or not. This would be a more general option and would help us to avoid cluttering the API for future changes.

File: autowrap/Utils.py [hunk 1]
```diff
@@ -83,6 +83,8 @@ def compile_and_import(name, source_files, include_dirs=None, **kws):
     else:
         compile_args = ""
 
+    compile_args += ", '-std=c++11' "
```
Comment (hroest) [inline autowrap/Utils.py]: I think this needs to be a parameter, I dont think we can simply add this to all code...

Comment (poshul) [reply autowrap/Utils.py]: Okay, what should the default be c++03 or c++11?

Comment (hroest) [reply autowrap/Utils.py]: I would say default should be C++03 to make this backwards compatible and act the same way after an upgrade

**Comment by cbielow** [comment]
is this ready to merge? We somewhat need this to go forward with https://github.com/OpenMS/OpenMS/pull/4311

**Comment by cbielow** [comment]
any estimate yet on the timeline?

File: autowrap/Utils.py (hunk 1 above)
Comment (uweschmitt) [reply autowrap/Utils.py:86]: Yes, before we merge this should be fixed. Please introduce an extra flag argument for this function and also check on windows. On Win this modification will result in `compile_args = ", '-std=c++11'"` which might break the generated `setup.py` file.
I recommend to collect the compile args as a list of strings and finally use `", ".join(...)` to create the final string.

**Comment by uweschmitt** [comment]
I suggested a change https://github.com/uweschmitt/autowrap/pull/83#discussion_r446271766 which should be resolved before we can merge.

File: autowrap/Utils.py (hunk 1 above)
Comment (cbielow) [reply autowrap/Utils.py:86]: just out of curiosity: what config/platform does not support c++11 these days?

Comment (uweschmitt) [reply autowrap/Utils.py:86]: We don't know. C++11 is also not guaranteed to be downwards compatible (https://stackoverflow.com/questions/6399615)

**Comment by timosachsenberg** [comment]
@poshul I incoporated your changes in OpenMS here https://github.com/OpenMS/OpenMS/pull/5180 .
Uwe suggested a fix https://github.com/uweschmitt/autowrap/pull/83#discussion_r446271766 and this PR is a bit outdated with some conflicting files. Any chance to get this updated?

File: autowrap/Utils.py (hunk 1 above)
Comment (poshul) [reply autowrap/Utils.py]: Okay, I've added a check in compile_and_import to check for which std it is passed. I've added options for c++11 14 17 and 20

**Review by poshul — COMMENTED** [review]
I've updated this to match the current upstream and addressed the suggestion from Uwe above.

**Comment by timosachsenberg** [comment]
should we perform a check e.g. if c++03 then std option for smart_ptr is not allowed

File: autowrap/Utils.py (hunk 1 above)
Comment (poshul) [reply autowrap/Utils.py]: @uweschmitt looking back at the original code, it looks like I was incorrect and this isn't necessary. Are you okay with the current revision which just removes it?

Comment (uweschmitt) [reply autowrap/Utils.py]: 👍

**Review by timosachsenberg — APPROVED** [review]
looks good to me

**Comment by timosachsenberg** [comment]
Hi @uweschmitt can you merge this PR or are there changes left we need to do?

**Comment by uweschmitt** [comment]
Uploaded `0.22.3` to `pypi.org`.

**Pull request merged**